### PR TITLE
compaction_backlog_tracker: do not allow moving registered trackers

### LINKED ISSUE #15248: compaction_backlog_tracker move ctor does not transfer the tracker registration properly
Currently the compaction_backlog_tracker move ctor moves the other backlog manager into `this`:
https://github.com/scylladb/scylladb/blob/cf37ab96f4cf5e429faaa2fc256bb295b1924b3a/compaction/compaction_manager.cc#L2125

But if `other` was registered to this manager, we botch registration, as:
1. `this` is not registered to the moved manager
2. `other` will not deregister itself from the moved manager in its destructor since it's left with a null manager here

### PR DESCRIPTION
**Pull request by bhalevy**
Currently, the moved-object's manager pointer is moved into the
constructed object, but without fixing the registration to
point to the moved-to object, causing #15248.

Although we could properly move the registration from
the moved-from object to the moved-to one, it is simpler
to just disallow moving a registered tracker, since it's
not needed anywhere. This way we just don't need to mess
with the trackers' registration.

The move-assignment operator has a similar problem,
therefore it is deleted in this series, and the function is
renamed to `transfer_backlog` that just doesn't deal with the
moved-from registration.  This is safe since it's only used internally
by the compaction manager.

Fixes #15248

### CODE REVIEW

File: compaction/compaction_manager.cc [hunk 1]
```diff
@@ -2209,7 +2205,7 @@ compaction_backlog_manager::~compaction_backlog_manager() {
 
 void compaction_manager::register_backlog_tracker(table_state& t, compaction_backlog_tracker new_backlog_tracker) {
     auto& cs = get_compaction_state(&t);
-    cs.backlog_tracker = std::move(new_backlog_tracker);
+    cs.backlog_tracker.transfer_backlog(std::move(new_backlog_tracker));
     register_backlog_tracker(cs.backlog_tracker);
```
Comment (raphaelsc) [inline compaction/compaction_manager.cc:2212]: what if we make backlog tracker optional and just rely on compaction_backlog_tracker's move ctor?

then we can delete move_assignment altogether
```

    compaction_state cs;

    compaction_backlog_tracker new_backlog_tracker;

    cs.backlog_tracker.emplace(std::move(new_backlog_tracker)); // this invokes move ctor.
```

Comment (bhalevy) [reply compaction/compaction_manager.cc:2212]: I think that should work

**Comment by bhalevy** [comment]
In v2 ([7ca91d7](https://github.com/scylladb/scylladb/commit/7ca91d719c8a5b1e211e24a4f9638b68783b8f7e)):
- compaction_state: store backlog_track in std::optional

**Review by raphaelsc — APPROVED** [review]

**Comment by bhalevy** [comment]
@scylladb/scylla-maint please merge

**Pull request merged**